src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/transforms/utils.ts

```diff
@@ -0,0 +1,314 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+import type { SavedObjectReference } from '@kbn/core-saved-objects-common/src/server_types';
+import type {
+  FormBasedLayer,
+  FormBasedPersistedState,
+  GenericIndexPatternColumn,
+  PersistedIndexPatternLayer,
+} from '@kbn/lens-plugin/public';
+import type {
+  TextBasedLayer,
+  TextBasedLayerColumn,
+  TextBasedPersistedState,
+} from '@kbn/lens-plugin/public/datasources/form_based/esql_layer/types';
+import type { AggregateQuery } from '@kbn/es-query';
+import { getIndexPatternFromESQLQuery } from '@kbn/esql-utils';
+import type { LensAttributes, LensDatatableDataset } from '../types';
+import type { LensApiState } from '../schema';
+import { fromBucketLensStateToAPI } from './columns/buckets';
+import { getMetricApiColumnFromLensState } from './columns/metric';
+import type { AnyLensStateColumn } from './columns/types';
+
+type DataSourceStateLayer =
+  | FormBasedPersistedState['layers'] // metric chart can return 2 layers (one for the metric and one for the trendline)
+  | PersistedIndexPatternLayer
+  | TextBasedPersistedState['layers'][0];
+
+export const getDefaultReferences = (
+  index: string,
+  dataLayerId: string
+): SavedObjectReference[] => {
+  return [
+    {
+      type: 'index-pattern',
+      id: index,
+      name: `indexpattern-datasource-layer-${dataLayerId}`,
+    },
+  ];
+};
+
+/**
+ * given Lens State layer and column id, returns the corresponding Lens API operation
+ * @param columnId
+ * @param layer
+ * @returns
+ */
+export const operationFromColumn = (columnId: string, layer: FormBasedLayer) => {
+  const column = layer.columns[columnId];
+  if (!column) {
+    return undefined;
+  }
+  // map columns to array of { column, id }
+  const columnMap = Object.entries(layer.columns).map(([id, c]) => ({
+    column: c as AnyLensStateColumn,
+    id,
+  }));
+  if (['terms', 'filters', 'ranges', 'date_range'].includes(column.operationType)) {
+    return fromBucketLensStateToAPI(column, columnMap);
```

Comment:
@dej611 this throws type errors and the only way i am able to get it to work is to remove all but the generic type definition for this method